zebra-consensus/src/block/check.rs

```diff
@@ -51,6 +57,41 @@ pub fn is_equihash_solution_valid(header: &Header) -> Result<(), equihash::Error
 /// accepted." [§7.5][7.5]
 ///
 /// [7.5]: https://zips.z.cash/protocol/protocol.pdf#blockheader
-pub fn is_time_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
+pub fn time_is_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
     header.is_time_valid_at(now)
 }
+
+/// [3.9]: https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts
+pub fn subsidy_is_correct(network: Network, block: &Block) -> Result<(), BoxError> {
+    let height = block
+        .coinbase_height()
+        .expect("always called on blocks with a coinbase height");
+
+    let coinbase = block.transactions.get(0).ok_or("no coinbase transaction")?;
+    let outputs = coinbase.outputs();
+
+    let canopy_height = Canopy.activation_height(network).ok_or("no canopy")?;
+    if height >= canopy_height {
+        panic!("Can't validate Canopy yet");
+    }
+
+    // validate founders reward and miner subsidy
+    if height > Height(0) && height < canopy_height {
+        let block_subsidy = subsidy::block_subsidy(height, network)?;
+        let miner_subsidy = subsidy::miner_subsidy(height, network)?;
+        let founders_reward = (block_subsidy / 5)?;
+
+        let values = || {
+            outputs
+                .iter()
+                .map(|o| o.value.try_into().expect("value will be a valid amount"))
+        };
+        if values()
+            .any(|value: Amount<NonNegative>| value == founders_reward && value != miner_subsidy)
+            && values().any(|value| value == miner_subsidy && value != founders_reward)
```

Comment:
It looks like my previous comments disappeared, because this code was modified.

A. Let's use the `founders_reward` function here?

B. These checks aren't correct, the correct checks are:

1. the exact founders reward value must be sent as a single output to the correct address
2. the sum of the coinbase transaction outputs must be less than or equal to the block subsidy plus transaction fees

So we're still missing the address check, and the sum of the coinbase transaction outputs check.

```suggestion
        let fr = founders_reward(height, network)?;

        let values = || {
            outputs
                .iter()
                .map(|o| o.value.try_into().expect("value will be a valid amount"))
        };
        if values()
            .any(|value: Amount<NonNegative>| value == fr)
```

Reply:
Based on our conversation on Discord, we want to check the miner fees in a future PR. So we can't do this check yet:
> 2. the sum of the coinbase transaction outputs must be less than or equal to the block subsidy plus transaction fees